### PR TITLE
(Feature/Change?): Real-time tree placer & select multiple treetypes

### PR DESCRIPTION
**Pull request by KajiKaede**
## Motivation / Problem
While decidedly functional, the current tree placement tool in regular games is... Dated. It's only fit for placing singular trees at a time, or using it's square-dragging functionality. Suitable for improving the local town authority's opinion of you, but... Also an incredible eyesore due to it's unnatural, perfectly square shape. OpenTTD is already host to the Scenario Editor's improved tree placer, but it is locked within that mode, and not programmed for multiplayer use.

## Description
This PR is meant to resolve that last circumstance. It replaces the in-game tree placer tool mechanics with those of the scenario editor. When holding down the mouse button, a tree is placed every time a user's cursor is dragged over a new tile, as well as allowing the user to select the options that randomly distribute trees in a radius.

...I was also working on allowing the user to select multiple trees at the same time, since I had to touch the same parts of the code. So that's in here too.

**Notes:**
- In multiplayer, due to the high risk of excessive commands, the tree placer limits itself to being ran four times. 
  - Any additional trees a user intends to plant are stored locally. Once the server has ran the command, another command is allowed to be sent, including all of the trees the user intended to place. 
  - While not as smooth as singleplayer, my testing seems to indicate the current method is unintrusive to the rest of the code, as well as being simple and tidy. It likely won't matter to most end users, and the results are still - in my opinion - a significant improvement over the current implementation in terms of visual feedback.

## Limitations
- This is my first C++ project, thus is especially vulnerable to mistakes due to lack of awareness or forgetfulness. 
- Worked in a local multiplayer testing scenario, but may need testing for potential edge cases or issues with long-distance/unstable connections.
- OpenTTD has previously had no brush-style tools like this. Every tile movement is a new cost measurement based on the new tiles placed. The newest cost popup includes the sum of all other, previously-placed popups. Due to the multiple popups being generated, visuals may be noisy, and a new visual effect for a repositionable cost counter may help make this smoother. I've hesitated to implement it thus far for the sake of avoiding intrusiveness in areas beyond this specific PR.
- This new functionality completely replaces the old tree placer functionality, whereas it may be more fit as an optional feature. While it should be easy to restore this functionality, I wanted to await discussion, especially as I'm not familiar yet with the process of implementing new settings.

### CODE REVIEW

File: src/tree_cmd.cpp [hunk 1]
```diff
@@ -805,6 +859,110 @@ CommandCost CmdPlantTree(DoCommandFlags flags, TileIndex end_tile, TileIndex sta
...
+		buffer.Send_uint8(t.second.type);
+		buffer.Send_uint8(t.second.count);
+	}
+}
+
+bool TreeSyncCmdData::Deserialise(DeserialisationBuffer &buffer, StringValidationSettings default_string_validation)
+{
+	this->calling_client = (ClientID)buffer.Recv_uint32();
+	uint32_t size = buffer.Recv_uint32();
+	if (!buffer.CanRecvBytes(size * 4)) return false;
```
Comment (JGRennison) [inline src/tree_cmd.cpp:926]: This looks like it should be `size * 6`.

File: src/tree_gui.cpp [hunk 2]
```diff
@@ -208,38 +266,36 @@ class BuildTreesWindow : public Window
...
-				Command<CMD_PLANT_TREE>::Post(tile, tile, this->tree_to_plant, false);
-			} else {
-				this->DoPlantForest(tile);
-			}
+		/* Test if the user is holding down the mouse button. base.OnPlaceDrag runs this at least once after user has let go, leading to an odd tree duplicating effect if we don't test. */
+		if (_left_button_down) {
+			this->DoPlantForest(tile);
+			SendSyncTrees();
+		}
+		if (_tree_placer_memory.size() >= 128 ) {
```
Comment (JGRennison) [inline src/tree_gui.cpp:281]: Is there meant to be something in this `if`?

File: src/tree_cmd.cpp [hunk 3]
```diff
@@ -805,6 +859,110 @@ CommandCost CmdPlantTree(DoCommandFlags flags, TileIndex end_tile, TileIndex sta
...
+	}
+	return true;
+}
+
+void TreeSyncCmdData::FormatDebugSummary(format_target &output) const
+{
+	output.format("Size {}", this->sync_data.size());
+}
+
+static int last_cmd_limit_tick_count = 0;
```
Comment (JGRennison) [inline src/tree_cmd.cpp:941]: This doesn't seem to be used anywhere

File: src/tree_cmd.cpp [hunk 4]
```diff
@@ -805,6 +859,110 @@ CommandCost CmdPlantTree(DoCommandFlags flags, TileIndex end_tile, TileIndex sta
...
+/**
+ * Sync trees, sent when a client is using the Tree Placer.
+ * @param flags type of operation
+ * @param cmd_data stuff
+ * @return the cost of this operation or an error
+ */
+CommandCost CmdSyncTrees(DoCommandFlags flags, const TreeSyncCmdData &cmd_data)
+{
+	if (_networking && cmd_data.calling_client == _network_own_client_id && flags.Test(DoCommandFlag::Execute)) {
+		_tree_placer_cmd_count--;
```
Comment (JGRennison) [inline src/tree_cmd.cpp:873]: This looks like it will result in it going negative, as ResetToolData clears it to 0, and then the commands from the server will presumably be arriving later.

File: src/tree_gui.cpp [hunk 5]
```diff
@@ -76,34 +82,47 @@ class BuildTreesWindow : public Window
 		PM_FOREST_LG,
 	};
 
-	int tree_to_plant = -1; ///< Tree number to plant, \c TREE_INVALID for a random tree.
+	int tree_types_base = _tree_base_by_landscape[to_underlying(_settings_game.game_creation.landscape)];
+	int tree_types_count = _tree_count_by_landscape[to_underlying(_settings_game.game_creation.landscape)];
+	std::set<TreeType> trees_to_plant = {}; /// < Container with every TreeType selected by the user.
```
Comment (JGRennison) [inline src/tree_gui.cpp:87]: This does seem like it could be a bitset type and avoid the overhead of std::set (this is not that important in the big scheme of things though).

File: src/tree_cmd.cpp [hunk 6]
```diff
@@ -805,6 +859,110 @@ CommandCost CmdPlantTree(DoCommandFlags flags, TileIndex end_tile, TileIndex sta
...
+	output.format("Size {}", this->sync_data.size());
+}
+
+static int last_cmd_limit_tick_count = 0;
+void SendSyncTrees()
+{
+	if (_tree_placer_memory.size() == 0) return;
+	if (_networking) {
+		if (_tree_placer_cmd_count > 4) return;
+		_tree_placer_cmd_count++;
```
Comment (JGRennison) [inline src/tree_cmd.cpp:947]: If the DoCommandP fails for whatever reason this might end up stuck at a positive value even with nothing in flight, but this might not any real problem in practice.

File: src/tree_cmd.h [hunk 7]
```diff
@@ -11,7 +11,22 @@
...
 #include "command_type.h"
+#include <map>
 
-DEF_CMD_TUPLE(CMD_PLANT_TREE, CmdPlantTree, CMD_AUTO, CMDT_LANDSCAPE_CONSTRUCTION, CmdDataT<TileIndex, uint8_t, bool>)
+struct TreePlacerData {
+	uint8_t type; uint8_t count;
+};
+
+struct TreeSyncCmdData final : public CommandPayloadSerialisable<TreeSyncCmdData> {
+	ClientID calling_client = INVALID_CLIENT_ID; // Client using this command.
```
Comment (JGRennison) [inline src/tree_cmd.h:21]: For tidyness, this does seem like it ought to use the CMD_CLIENT_ID flag on the command, probably with a GetClientIDField method, but that's just a minor thing.

**Comment by KajiKaede** [comment]
Seen everything mentioned. Some of it seems to be carry-overs from older iterations of the code or things I was concerned about but forgot to mention. Very much appreciated. I'll go over all of these in a moment and tidy up what I can.

File: src/tree_gui.cpp (hunk 5 above)
Comment (KajiKaede) [reply src/tree_gui.cpp:87]: I think this was a case where I was also considering the potential ramifications of a future implementation of custom tree GRFs, though I'm not sure of the demand for such a feature, if a set would even be necessary in the first place for that case either. Bitset seems like a more sensible solution in the meantime, though.

File: src/tree_cmd.h (hunk 7 above)
Comment (KajiKaede) [reply src/tree_cmd.h:21]: Think I might leave this one up to someone else. There doesn't appear to be any existing examples as a cross-reference for tidying this up, and my attempts would likely take enough questions that it'd just be better done by someone who can do it correctly with confidence first time.

**Comment by KajiKaede** [comment]
Didn't mean to put that last commit here yet. Seem to be getting errors on launch I need to fix.

**Comment by JGRennison** [comment]
Normal tree placing ought to still be available, I think, but I can take care of that.
Cost estimation also doesn't seem to work from a UI PoV (not that this is all that useful).
The command layer is firing every tick, I will see what can done about that though.
A cost message on every tick is indeed rather noisy, and cumulative values is a bit of a departure from how they normally work. Adjustments to the command layer may well make this less of an issue though.
I will do some more tinkering.

**Comment by KajiKaede** [comment]
I may very well have messed something up with my new changes in terms of commands, I had forgotten about testing multiplayer with the new changes.
Having the option for users to use the original behaviour is an option I was thinking would be preferable as well. Restoring it should be pretty easy on my end if I'm to keep working on this.
Agreed with the cumulative values, though this entire tool strays from current OpenTTD norms of design expectations. I feel it would be more intuitive were it not for the previous cost messages lingering, but it's very much a question of design that I'm not qualified to make with certainty.

Let me know if there's any of this I should work on. Don't want to intrude if you're also working on any of these issues.

**Comment by KajiKaede** [comment]
Aye, think that's a better call for now.
I personally like the idea of being able to drag the trees, but original behaviour on 'normal' matches with the current scenario editor, plus less options bloat and - most important of all - less work.

**Comment by KajiKaede** [comment]
Oh. TIL EnumBitSet is a thing. Jeez I've still got a lot to learn.

**Comment by KajiKaede** [comment]
Hope it's not a bother for me to keep submitting my work in future. My work is likely to keep needing some level of revision/tidying for a while (though seeing these commits is very useful for identifying mistakes/features I can take note of for the future). Thus, I want to check if it's fine to keep working on features and submitting PRs that may contain issues like this. My guess is that it's fine, but I want to be courteous and make sure I'm not creating any manner of inconvenience or obligational pressure by my work. (I already know that not every proposed idea will be accepted. I just want to ensure I'm not causing any issues by offering code that may need improving/fixing.)
